toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -0,0 +1,177 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/builtin_function_kind.h"
+
+#include <utility>
+
+#include "toolchain/sem_ir/file.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::SemIR {
+
+namespace BuiltinFunctionInfo {
```

Comment:
Why not use an anonymous namespace, particularly given you have one on line 21?

If this is so that you can do `constexpr BuiltinInfo None = {"", nullptr};`, I wonder if there would be easier to understand ways to do this. For example, had you considered something like creating a std::array of BuiltinInfo objects, and indexing into that? It could be assigned by index at construction in order to avoid ambiguity about ordering, which I assume was one thing you were trying to avoid with the current approach.

Reply:
My one last comment here would be, and up to you, but you could take the individual names and put them in an array for a macro, then use the single array in place of both the StringSwitch and validate function. Even name() could be replaced with that, I think.

Reply:
I've looked into this. `EnumBase` really wants its own names array, so we'd end up storing the names twice in that direction: once in the `BuiltinFunctionInfo` table and again in the names table. This is probably worth looking into again if we add another field into `BuiltinFunctionInfo`, but I'm inclined to not do this for now.